lib/dfu_target_uart_host/dfu_target_uart_host.c

```diff
@@ -0,0 +1,319 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ * Copyright (c) 2022 INTELLINIUM <giuliano.franchetto@intellinium.com>
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <drivers/uart.h>
+#include <string.h>
+#include <dfu_target_uart_host.h>
+#include <dfu/dfu_target_mcuboot.h>
+#include <dfu/dfu_target.h>
+#include <zephyr/logging/log.h>
+#include <stdnoreturn.h>
+#include <dfu_target_uart_encode.h>
+#include <dfu_target_uart_decode.h>
+#include <zephyr/sys/ring_buffer.h>
+
+LOG_MODULE_REGISTER(dfu_t_uart_host, CONFIG_DFU_TARGET_UART_HOST_LOG_LEVEL);
+
+K_THREAD_STACK_DEFINE(uart_handler_stack,
+		      CONFIG_DFU_TARGET_UART_HOST_HANDLER_STACK_SIZE);
+static struct k_thread uart_handler_thread;
+static k_tid_t tid;
+static uint8_t in_buf[CONFIG_DFU_TARGET_UART_HOST_IN_BUFFER_SIZE];
+
+static struct dfu_target_uart_host_ctx {
+	/* The UART device used for the DFU */
+	const struct device *uart_dev;
+
+	/* The ring buffer used to store received from the UART bus */
+	struct ring_buf rx_rb;
+
+	/* Semaphore used to wait data from UART bus */
+	struct k_sem lte_uart_sem;
+
+	size_t file_size;
+	uint8_t image_num;
+	bool download_success;
+
+	/* Chunk of upgrade application to write */
+	const uint8_t *to_write;
+	size_t to_write_len;
+
+	/* The output buffer that will be sent to the host core */
+	uint8_t out_buffer[CONFIG_DFU_TARGET_UART_HOST_OUT_BUFFER_SIZE];
+	size_t out_buf_cursor;
+} context;
+
+static void dfu_target_uart_host_send_out_buffer(void)
+{
+	LOG_DBG("Sending %zu bytes", context.out_buf_cursor);
+
+	for (int i = 0; i < context.out_buf_cursor; i++) {
+		uart_poll_out(context.uart_dev, context.out_buffer[i]);
+	}
+
+	context.out_buf_cursor = 0;
+}
+
+static void dfu_target_uart_host_send_error(int func, int error)
+{
+	int err;
+
+	const struct DFU_UART_target_RemoteResp fnc = {
+		.type_choice = func,
+		.error = error
+	};
+
+	err = cbor_encode_DFU_UART_target_RemoteResp(
+		context.out_buffer,
+		ARRAY_SIZE(context.out_buffer),
+		&fnc,
+		&context.out_buf_cursor);
+	if (err) {
+		LOG_ERR("Could not encode response, error %d", err);
+		return;
+	}
+
+	if (context.out_buf_cursor == 0) {
+		LOG_DBG("Could not send error, no data in buffer");
+		return;
+	}
+
+	dfu_target_uart_host_send_out_buffer();
+}
+
+static void dfu_target_uart_host_send_offset(int error, size_t offset)
+{
+	int err;
+	int val = error == 0 ? (int) offset : error;
```

Comment:
Not following this, there is a dedicated send error function, why is this being used for both an error if negative or offset if positive?